lib/core/async/event_synchronizer.dart

```diff
@@ -0,0 +1,145 @@
+/// it returns `null` or `false`.
+typedef _Subscriber = Future<bool?> Function(dynamic);
+
+/// A class which helps managing asynchronuous, callback-based communnication.
+///
+/// [EventSynchronizer] allows to request, expect and exchange messages between two independent
+/// endpoints, e.g. web sockets, isolates, ...
+///
+/// ```dart
+/// // Some I/O client which sends and receives messages.
+/// final client = IOCLient()
+///
+/// // The synchronizer is initialized with the send function of the client.
+/// final synchronizer = EventSynchronizer(client.send);
+///
+/// // We need to pass each message from the client to the synchronizer.
+/// client.onMessage = (dynamic message) => synchronizer.handleMessage(message);
+///
+/// // Simple messages are sent using the request function.
+/// await synchronizer.request('ANYTHING');
+///
+/// // We can receive the String response directly using an identity function.
+/// final String response = await synchronizer.expect(String, (String resp) => resp);
+///
+/// // If we are witing for a specific response and can ignore other messages, we can use keepAlive.
+/// // It will wait with completion until we receive "THAT'S IT".
+/// // The return value in this case is `true`.
+/// await synchronizer.expect(String, (String resp) => resp == "THAT'S IT", keepAlive: true);
+///
+/// // Using exhange, we can reallize a ping-pong communication.
+/// // Use drain to await the stream completion.
+/// synchronizer.exchange(someStream, String, (String req, String resp) => resp).listen((String resp) {
+///  print(resp);
+/// });
+/// ```
+class EventSynchronizer {
+  /// The sener function used to dispatch messages.
+  final Sender _sender;
+
+  /// A map of message subscriptions identified by a random UUID.
+  final Map<String, _Subscriber> _subscriptions = {};
+
+  /// UUID generator.
+  final _uuid = Uuid();
+
+  /// A time offset required if the I/O responses might be faster than the code which attaches
+  /// a listener ([expect]) for them.
+  ///
+  /// The duration is a minimum time which is required for each [handleMessage] call.
+  /// Use with caution!
+  final Duration messageOffset;
+
+  EventSynchronizer(this._sender, [this.messageOffset = Duration.zero]);
+
+  /// Message handler which should be called for each incoming [message] without pre-filtering.
+  Future<void> handleMessage(dynamic message) async {
+    await Future.delayed(messageOffset);
+
+    List<String> toKeep = [];
+
+    for (final entry in _subscriptions.entries) {
+      final id = entry.key;
+      final subscription = entry.value;
+
+      if (await subscription(message) == true) {
+        toKeep.add(id);
+      }
+    }
+
+    _subscriptions.removeWhere((key, value) => !toKeep.contains(key));
+  }
+
+  /// Sends a [message] using the [_sender] function.
+  ///
+  /// If [message] is a [Stream] each item will be sent seperately.
+  Future<void> request(dynamic message) async {
```

Comment:
Yes we do, each time we call it to be accurate. Assume the send method does some async work which requires some time depending on the message. Not awaiting could lead to this:

sender code
```dart
request(5); // takes a long time to send because we need to read the content of file 5 or something
request(6); // that's much faster
```

we'd receive 6 and then 5, even though we sent 5 before 6

Reply:
To avoid this, we could simply use [synchronized](https://pub.dev/packages/synchronized) which we already use in `encryptionMiddleware`

Reply:
Looks like a lot of overhead when a simple await does the job too :P